File: app/backend/src/services/logger.ts
import fs from 'fs';
import path from 'path';
import { LOG_DIR } from '../config';
import { LogEntry } from '../types';

class Logger {
  private logFile: string;
  private auditFile: string;

  constructor() {
    if (!fs.existsSync(LOG_DIR)) {
      fs.mkdirSync(LOG_DIR, { recursive: true });
    }
    this.logFile = path.join(LOG_DIR, 'events.jsonl');
    this.auditFile = path.join(LOG_DIR, 'audit.jsonl');
  }

  private write(filePath: string, entry: Omit<LogEntry, 'timestamp'>): void {
    const fullEntry: LogEntry = {
      ...entry,
      timestamp: new Date().toISOString(),
    };
    try {
      fs.appendFileSync(filePath, JSON.stringify(fullEntry) + '\n', 'utf-8');
    } catch (err) {
      console.error('[Logger] Failed to write log:', err);
    }
  }

  info(message: string, metadata?: Record<string, unknown>): void {
    this.write(this.logFile, { level: 'info', message, metadata });
  }

  error(message: string, metadata?: Record<string, unknown>): void {
    this.write(this.logFile, { level: 'error', message, metadata });
  }

  audit(message: string, metadata?: Record<string, unknown>): void {
    this.write(this.auditFile, { level: 'audit', message, metadata });
  }
}

export const defaultLogger = new Logger();
export default Logger;

File: src/logger/logger.py
import json
from datetime import datetime
from pathlib import Path
from typing import Any
from ..config import LOG_DIR


class Logger:
    def __init__(self, log_dir: Path | str = LOG_DIR):
        self.log_dir = Path(log_dir)
        self.log_dir.mkdir(parents=True, exist_ok=True)
        self.log_file = self.log_dir / "events.jsonl"
        self.audit_file = self.log_dir / "audit.jsonl"

    def _write(self, path: Path, obj: dict):
        obj["timestamp"] = datetime.utcnow().isoformat() + "Z"
        with open(path, "a", encoding="utf-8") as f:
            f.write(json.dumps(obj, ensure_ascii=False) + "\n")

    def info(self, message: str, **metadata: Any):
        self._write(self.log_file, {"level": "info", "message": message, "metadata": metadata})

    def error(self, message: str, **metadata: Any):
        self._write(self.log_file, {"level": "error", "message": message, "metadata": metadata})

    def audit(self, message: str, **metadata: Any):
        self._write(self.audit_file, {"level": "audit", "message": message, "metadata": metadata})


default_logger = Logger()
